network/phonebook.go

```diff
@@ -86,43 +153,66 @@ func (p *ThreadsafePhonebook) ExtendPeerList(more []string) {
 	defer p.lock.Unlock()
 	// TODO: if this gets bad because p.addrs gets long, replace storage with a map[string]bool
 	for _, addr := range more {
-		found := false
-		for _, oaddr := range p.addrs {
-			if addr == oaddr {
-				found = true
-				break
-			}
-		}
-		if !found {
-			p.addrs = append(p.addrs, addr)
+		if _, has := p.entries[addr]; has {
+			continue
 		}
+		p.entries[addr] = phonebookData{}
 	}
 }
 
 // Length returns the number of addrs contained
 func (p *ThreadsafePhonebook) Length() int {
 	p.lock.RLock()
 	defer p.lock.RUnlock()
-	return len(p.addrs)
+	return len(p.entries)
 }
 
 // ReplacePeerList replaces set of addresses with that passed in.
 func (p *ThreadsafePhonebook) ReplacePeerList(they []string) {
 	p.lock.Lock()
 	defer p.lock.Unlock()
-	p.addrs = make([]string, len(they))
-	copy(p.addrs, they)
+	p.entries.ReplacePeerList(they)
+}
+
+// MergePeerList merges a set of addresses with that passed in.
```

Comment:
This is what I wish ReplacePeerList did. Why should ReplacePeerList forget the metadata associated with an address that was there before and is there in the new list?

Reply:
Good point; I'll merge the two.